### PR TITLE
Some improvements and bug fixes

### PR DESCRIPTION
**Pull request by L11R**
Consider it as a request for comments, not real PR that should be merged.

There is some points:

0. There is go modules, why do we still use `vendor`? All packages are cached by default by Google, you can also use own cache proxy at your organization to make builds reproducible. 
1. We use Genesys Web Services, and it support only Cookie-based auth, so I have to bypass own HTTP client with own cookie jar.
2. You have bug here: https://github.com/sigmavirus24/gobayeux/blob/master/client.go#L31
```go
func NewClient(serverAddress string, logger logrus.FieldLogger) (*Client, error) {
	bc, err := NewBayeuxClient(nil, serverAddress, logger)
	if err != nil {
		return nil, err
	}
	if logger == nil {
		logger := logrus.New()
		logger.SetLevel(logrus.PanicLevel) // You cannot call logger.SetLevel because logrus.FieldLogger interface HAS NOT this method
	}
```
Fix:
```go
if logger == nil {
	l := logrus.New()
	l.SetLevel(logrus.PanicLevel)
	logger = l
}
```
3. Your current high-level Client is not configurable at all, so added opts pattern approach.
4. I don't know why you made `Data` field as binary representation as default. It's optional, Genesys sends just own JSON structures here. My patch:
```go
	// Data contains the event information and optionally could contain
	// a binary representation of the data
	// See also https://docs.cometd.org/current/reference/#_data and
	// https://docs.cometd.org/current/reference/#_concepts_binary_data
	//
	// There is the MessageData type can be used for binary data
	Data json.RawMessage `json:"data,omitempty"`
```
5. I don't know how Saleforce backend works, but in case of Genesys meta-message about success come last, not first. So instead of doing like that:
```go
	response, err := b.parseResponse(resp)
	if err != nil {
		logger.WithError(err).Debug("error parsing response")
		return response, err
	}

	if !response[0].Successful {
		return response, errors.New("connect request was not successful")
	}
```
I am doing like this:
```go
	response, err := b.parseResponse(resp)
	if err != nil {
		logger.WithError(err).Debug("error parsing response")
		return response, err
	}

	for _, m := range response {
		if m.Channel == MetaConnect && !m.Successful {
			return response, errors.New("connect request was not successful")
		}
	}
```

### CODE REVIEW

File: message.go [hunk 1]
```diff
@@ -27,9 +28,13 @@ type Message struct {
...
-	// Data contains the binary represnetation of the data
-	// See also https://docs.cometd.org/current/reference/#_data
-	Data *MessageData `json:"data,omitempty"`
+	// Data contains the event information and optionally could contain
+	// a binary representation of the data
+	// See also https://docs.cometd.org/current/reference/#_data and
+	// https://docs.cometd.org/current/reference/#_concepts_binary_data
+	//
+	// There is the MessageData type can be used for binary data
+	Data json.RawMessage `json:"data,omitempty"`
```
Comment (sigmavirus24) [inline message.go:37]: I think I misread the docs in this particular case. I see two examples one where `data` at this level is just a string, and another where it has a nested object that has its own `data` field. I'm comfortable getting rid of `MessageData` altogether

**Comment by L11R** [comment]
```go
func NewClient(serverAddress string, logger logrus.FieldLogger) (*Client, error) {
	bc, err := NewBayeuxClient(nil, serverAddress, logger)
	if err != nil {
		return nil, err
	}
	if logger == nil {
		logger := logrus.New()
		logger.SetLevel(logrus.PanicLevel)
	}```

oh, here I gave the incorrect explanation of bug, the problem is that you create the new variable with the same name but `*Client` will still have nil `logger` from outer scope. It should be definitely fixed.

File: message.go (hunk 1 above)
Comment (L11R) [reply message.go:37]: @sigmavirus24 so just json.RawMessage is okay and MessageData could be entirely removed?

**Comment by L11R** [comment]
Would be nice to here you opinion about other points. We could make v2 version to make it compatible with previous version, but anyway I do not see any tags currently.

**Comment by L11R** [comment]
@sigmavirus24 btw what is the point of explicit setting cookies? I mean this (there are more examples):
https://github.com/sigmavirus24/gobayeux/blob/master/bayeux_client.go#L282-L291

Golang HTTP client automatically sets them out of the box:
https://github.com/golang/go/blob/master/src/net/http/client.go#L180-L184

**Comment by sigmavirus24** [comment]
> Would be nice to hear you opinion about other points. We could make v2 version to make it compatible with previous version, but anyway I do not see any tags currently.

I'll get to it. This is mostly a for-fun/learning more about Go project for me than a serious open source effort. 

> btw what is the point of explicit setting cookies?

So CometD is kind of free-form. That means you might hit a certain endpoint, but there are other endpoints you might hit as part of it. So for a `/meta/handshake` some clients will append `/meta/handshake` to the "base" URL. And some server implementations return a cookie that then gets scoped to `https://example.com/cometd/meta/handshake` so it's common among cometd/bayeux clients to set those cookies for the domain rather than have default cookie jar logic apply as it could cause problems with server implementations expecting to see that cookie on all subsequent requests.

It's annoying and unspecified behaviour but as we know with HTTP, just because it's in the RFC doesn't mean it isn't ridiculously widespread and commonplace. I know based on other clients I've used/modified/written, for example, that Salesforce relies on clients to do this and so that was a motivating reason behind including that logic here. Node JS clients (primarily Faye) do this for all implementations with no negative effects it seems

**Comment by L11R** [comment]
@sigmavirus24 I got the idea behind this cookie logic and personally faced with such behavior in other cases, but your code just repeats default behavior, doesn't it?

Your code:
```go
for _, cookie := range b.client.Jar.Cookies(b.serverAddress) {
		req.AddCookie(cookie)
	}

	resp, err := b.client.Do(req)
	if err != nil {
		return nil, err
	}

	b.client.Jar.SetCookies(b.serverAddress, resp.Cookies())
	return resp, nil
```

Go's code:
```go
if c.Jar != nil {
		for _, cookie := range c.Jar.Cookies(req.URL) {
			req.AddCookie(cookie)
		}
	}
	resp, didTimeout, err = send(req, c.transport(), deadline)
	if err != nil {
		return nil, didTimeout, err
	}
	if c.Jar != nil {
		if rc := resp.Cookies(); len(rc) > 0 {
			c.Jar.SetCookies(req.URL, rc)
		}
	}
```

AFAIK your `serversAddress` is exactly their `req.URL`. Maybe I still doesn't understand something.

Also check my fork, I add README badges, fixed actions, typos in comments, etc. Of course I will replace L11R/gobayeux to sigmavirus24/gobayeux if you will consider to merge my changes completely.

**Comment by L11R** [comment]
So I squashed my small incremental commits into 1 and prepared the v1.0.0 release:
https://pkg.go.dev/github.com/L11R/gobayeux

Would be happy to upstream it.

**Comment by sigmavirus24** [comment]
> I got the idea behind this cookie logic and personally faced with such behavior in other cases, but your code just repeats default behavior, doesn't it?

It doesn't. The path attribute on the Cookie will prevent it from working appropriately. Right now we don't support appending the request path (e.g., `/meta/conect`, `/meta/handshake`) to the server address but that was functionality I was going to add. That changes the needs here

**Comment by L11R** [comment]
@sigmavirus24 by the way in case of Genesys it appends in other way:
There is a base URL: `https://genesys.company.org/api/v2/notifications`
And then it appends just `/handshake` or `/connect`, so it becomes:
- `https://genesys.company.org/api/v2/notifications/handshake`
- `https://genesys.company.org/api/v2/notifications/connect`

But it works perfectly even without that appending. I know that it appends only after researching it web frontend client.

**Comment by sigmavirus24** [comment]
I'm trying to make time to review this.  I'm inclined to accept it though since you've tested against a real service.  Could you make the namespace updates until then?

**Comment by L11R** [comment]
@sigmavirus24 sure, done!

File: message.go (hunk 1 above)
Comment (sigmavirus24) [reply message.go:43]: Yes, but I can do that in a follow-up here

**Comment by sigmavirus24** [comment]
I'm going to remove the MessageData and then create a v1.0.

I am confused about the `IDStorer` -> `IDStore` rename given I thought interfaces were supposed to be things like `Closer`, `Reader`, etc. but :shrug:

**Comment by sigmavirus24** [comment]
Done: https://github.com/sigmavirus24/gobayeux/releases/tag/v1.0.0

**Comment by L11R** [comment]
> I am confused about the `IDStorer` -> `IDStore` rename given I thought interfaces were supposed to be things like `Closer`, `Reader`, etc. but :shrug:

Well, I dont think there is some strict naming convention, Reader and Closer are at least existing nouns. Storer is non-existent english word, so I had a little bit confused by it and changed it.

AFAIK it's totally fine to just name interfaces using any nouns.

**Comment by sigmavirus24** [comment]
https://www.merriam-webster.com/dictionary/storer

**Comment by L11R** [comment]
@sigmavirus24 oh, okay, English is not my native, but it still sounds like an artificial word for me :)

**Comment by sigmavirus24** [comment]
Yeah, it's not common. And we've published v1.0 so we can't change it now :smile:

**Pull request merged**